README.md
# two-danode
Two Dimenisonal Autonomous Non-Linear Differential Equations Visualizer

See the book [Nonlinear Dynamics and Chaos](http://www.stevenstrogatz.com/books/nonlinear-dynamics-and-chaos-with-applications-to-physics-biology-chemistry-and-engineering) by Steven H. Strogatz

Amazing lectures by Prof. Strogatz on YouTube are [here](https://www.youtube.com/playlist?list=PLbN57C5Zdl6j_qJA-pARJnKsmROzPnO9V)

![alt text](https://farm5.staticflickr.com/4704/25792673807_32edd52ff5_z_d.jpg "Example")

## Description
This is a program that can take two functions F (x,y) and G (x,y) from the command line, representing F => &#7819; and G => &#7823; (The change in x and y respectively), and represent the resulting vectors in the [HSL color space](https://en.wikipedia.org/wiki/HSL_and_HSV). 

This isn't scientific, I thought it might be semi intuitive/cool to see the direction of vectors represented by the hue (the 'H' in HSL), such that unstable/stable nodes would be surrounded by a full rainbow. 

Vectors going right are red (0&#176;), then going counter-clockwise on the rainbow, yellow is 60&#176;, teal is 180&#176;, blue is 240&#176;, mageneta is 300&#176; and then back to red at 360&#176;. Magnitude of the vectors are represented by the intensity of the color up to white where the lightness (the 'L' in HSL) is 1 (normalized by the largest magnitude vector).

This can be illustrated with &#7819; = x and &#7823; = y, resulting in

![alt text](https://farm5.staticflickr.com/4746/39953334044_ba1714ba5b_m_d.jpg "Example")

Where right is red, left is teal and the origin is black (zero magnitude, 'L' = 0)
To get this image you can run 

```stack exec -- two-danode-exe -F "x" -G "y" -f ~/Desktop/cartesian.bmp``` 

or in polar coordinates 

```stack exec -- two-danode-exe -P -R "r" -T "0" -f ~/Desktop/polar.bmp``` 

with R and T a function of variables "r" and "theta".

## Installation and Usage
Install [Haskell Stack](https://docs.haskellstack.org/en/stable/README/), cd into directory ```stack install```

You have to run the binary with ```stack exec -- two-danode-exe```, this is because to eval command line arguments in Haskell I use the hint package which needs the environment variables for the location of the installed packages to interpret the equations and return the vector field.

See ```stack exec -- two-danode-exe -h``` for flags and usage. 

## Example 1
![alt text](https://farm5.staticflickr.com/4656/39953332894_1c5c8caf2b_k_d.jpg "Example")
```time stack exec -- two-danode-exe -H 5 -F "(1.5*sin (3*x*y -10))*(abs $ cos (1.75*x))**(abs $ 3 * cos (3*y)) + sin (3*x)" -G "sin (4.3*x) - cos (0.5*y)" -r "(4096,4096)"```

## Example 2
![alt text](https://farm5.staticflickr.com/4704/25792673807_a300fe656f_k_d.jpg "Example")
```time stack exec -- two-danode-exe -H 5 -r "(4096,4096)" -f 051.bmp -F "sin(2*y - 3*x)*exp(-abs (sin (3*x+1-2*y) - cos(x-3*y+1)))" -G "cos(2*y - 3*x)*exp(-abs (cos (3*x+1-2*y) - cos(x-3*y+1)))"```

## Example 3
![alt text](https://farm5.staticflickr.com/4611/25792677967_1afd9e8d9b_k_d.jpg "Example")
```time stack exec -- two-danode-exe -F "(cos (5*x*y))**sin (10*(x*y-1))*(0.5-x*y)" -G "0.1*(0.5-x*y)*10*cos (2-x*y*y)*(sin (5*y*x))" -H 4 -m 3000 -r "(4096,4096)" -f 022.bmp```

## Example 4
![alt text](https://farm5.staticflickr.com/4652/25792678497_4514f19b96_k_d.jpg "Example")
```time stack exec -- two-danode-exe -F "(cos (5*x))*(1-x)*(x-1)*(0.5-x)" -G "20*cos (1-x)*(sin (5*y))*(1-y)" -H 5 -m 30 -r "(4096,4096)" -f 020.bmp```

## Example 5
![alt text](https://farm5.staticflickr.com/4664/39953335404_98a586d91f_k_d.jpg "Example")
```time stack exec -- two-danode-exe -F "(cos (5*x*y))*(1-x*y)*sin (10*(x*y-1))*(0.5-x*y)" -G "0.1*(0.5-x*y)*10*cos (1-x*y*y)*(sin (5*y*x))*(1-y*x)" -H 4 -m 3000 -r "(4096,4096)" -f 021.bmp```

## More Examples 
More examples can be found on my [Flickr account](https://flic.kr/s/aHskxk4mB5)


ChangeLog.md
# Changelog for two-danode

## Unreleased changes


app/Main.hs
{-# LANGUAGE TemplateHaskell #-}

module Main where

import Field

import Control.Lens
import Control.Monad.Identity (runIdentity)

import Data.Array.Repa (Array,DIM2,DIM3,U,Z(..),(:.)(..),deepSeqArray)
import qualified Data.Array.Repa as R
import Data.Array.Repa.IO.BMP (writeImageToBMP)

import Data.Colour.RGBSpace
import Data.Colour.RGBSpace.HSL

import Data.Vector.Unboxed (Vector,Unbox)

import Data.Word (Word8)

import Linear

import Language.Haskell.Interpreter

import System.Console.GetOpt
import System.Environment (getArgs, getProgName)
import System.Exit
import System.IO

data Options =
  Options
  { _optFD :: FieldDescription Double
  , _optFile :: FilePath
  , _optLogMultiplier :: Maybe Double
  , _optPolar :: Bool
  , _optR :: String
  , _optT :: String
  , _optF :: String
  , _optG :: String
  } deriving Show

makeLenses ''Options

startOptions :: Options
startOptions =
  Options
  { _optFD = defaultFieldDescription
  , _optF = "y"
  , _optG = "-sin x"
  , _optPolar = False
  , _optR = "r*(1-r*r)"
  , _optT = "1"
  , _optLogMultiplier = Nothing
  , _optFile = "default.bmp"
  }

options :: [OptDescr (Options -> IO Options)]
options =
  [ Option "F" []
    (ReqArg
     (\arg opt -> pure $ optF .~ arg $ opt)
      "String")
    $ unlines $ ["Function String for 'x dot', F (x,y) = _"
                ,"Default: " ++ show (_optF startOptions)]
  , Option "G" []
    (ReqArg
     (\arg opt -> pure $ optG .~ arg $ opt)
      "String")
    $ unlines $ ["Function String for 'y dot', G (x,y) = _"
                ,"Default: " ++ show (_optG startOptions)]
  , Option "P" []
    (NoArg
     (\opt -> pure $ optPolar .~ True $ opt))
    $ unlines $ ["Set to use polar coordinates in terms of r and theta"
                ,"as r-dot and theta-dot"]
  , Option "R" []
    (ReqArg
     (\arg opt -> pure $ optR .~ arg $ opt)
      "String")
    $ unlines $ ["Function String for polar 'r dot', R (r,theta) = _"
                ,"Default: " ++ show (_optR startOptions)]
  , Option "T" []
    (ReqArg
     (\arg opt -> pure $ optT .~ arg $ opt)
      "String")
    $ unlines $ ["Function String for polar 'theta dot', T (r,theta) = _"
                ,"Default: " ++ show (_optT startOptions)]
  , Option "f" []
    (ReqArg
     (\arg opt -> pure $ optFile .~ arg $ opt)
      "File")
    $ unlines $ ["BMP file save name"
                ,"Default: " ++ show (_optFile startOptions)]
  , Option "a" ["aa"]
    (ReqArg
      (\arg opt -> pure $ optFD.aa .~ read arg $ opt)
      "Int")
    $ unlines $ ["Anti-Aliasing"
                ,"Default: " ++ show (startOptions ^. optFD.aa)]
  , Option "r" ["res"]
    (ReqArg
     (\arg opt -> pure $ optFD.res .~ read arg ^. tupleToV2 $ opt)
      "(Int,Int)")
    $ unlines $
    ["Resolution of output image, needs quotes in cmd line"
    ,"Default: " ++ show (startOptions ^. optFD.res.v2ToTuple.to show)]
  , Option "c" ["center"]
    (ReqArg
     (\arg opt -> pure $ optFD.center .~ read arg ^. tupleToV2 $ opt)
      "(Double,Double)")
    $ unlines $
    ["Center of field View, needs quotes in cmd line"
    ,"Default: " ++ show (startOptions ^. optFD.center.v2ToTuple.to show)]
  , Option "H" ["height"]
    (ReqArg
     (\arg opt -> pure $ optFD.h .~ read arg $ opt)
      "Double")
    $ unlines $ ["Height of field view"
                ,"Default: " ++ show (startOptions ^. optFD.h)]
  , Option "m" ["multiplier"]
    (ReqArg
     (\arg opt -> pure $ optLogMultiplier .~ Just (read arg)  $ opt)
      "Double")
    $ unlines $
    ["Multiplier of log function to convert colors,"
    ,"a higher number means less intensity variability "
    ,"between low and high magnitude vectors"
    ,"Default: " ++ show (startOptions ^. optLogMultiplier)]
  , Option "h" ["help"]
    (NoArg
      (\_ -> do
          prg <- getProgName
          hPutStrLn stderr $ usageInfo prg options
          exitWith ExitSuccess
      )
    ) "Show help"
  ]

getUsage :: IO String
getUsage =
  do
    prg <- getProgName
    return $ usageInfo prg options

createFunction :: Options -> String
createFunction opts
  | _optPolar opts =
    "(\\(V2 y x) -> let " ++
    "r = sqrt $ x*x + y*y" ++ "; " ++
    "theta = atan2 y x" ++ "; " ++
    "rdot = " ++ _optR opts ++ "; " ++
    "thetadot = " ++ _optT opts ++ "; " ++
    "xdot = rdot*cos theta - r*sin theta * thetadot" ++ "; " ++
    "ydot = rdot*sin theta + r*cos theta * thetadot" ++ "; " ++
    "in V2 ydot xdot)"
  | otherwise =
    "(\\(V2 y x) -> let " ++
    "xdot = " ++ _optF opts ++ "; " ++
    "ydot = " ++ _optG opts ++ "; " ++
    "in V2 ydot xdot)"

main :: IO ()
main =
  do
    args <- getArgs
    let (actions,_,_) = getOpt RequireOrder options args
    opts <- foldl (>>=) (return startOptions) actions
    let
      funcS = createFunction opts
      file = _optFile opts
      fd = _optFD opts
      dim = let V2 y x = _res fd
                aa' = _aa fd
            in Z :. y :. x :. aa'*aa' :: DIM3
      logMul = _optLogMultiplier opts
    result <-
      runInterpreter $
      do
        let command =
              unwords
              [ "toUnboxed $ buildFieldRepa"
              , "(" ++ show fd ++ ")"
              , funcS
              ]
        setImports [ "Data.Array.Repa"
                   , "Data.Vector.Unboxed"
                   , "Linear"
                   , "Prelude"
                   , "Field"
                   ]
        interpret command (as :: Vector (V2 Double))
    case result of
      Left err -> print err
      Right v ->
        writeImageToBMP file $ makeImage fd logMul $ R.fromUnboxed dim v

makeImage
  :: (Epsilon a, Unbox a, RealFrac a, Floating a, Ord a)
  => FieldDescription a
  -> Maybe a --logMul
  -> Array U DIM3 (V2 a)
  -> Array U DIM2 (Word8,Word8,Word8)
makeImage fd logMul vectorField =
  let
    aa' = _aa fd
    aaSq' = fromIntegral $ aa'*aa'
    maxV = runIdentity $ R.foldAllP max 0 $ R.map norm $ vectorField
    image' = runIdentity $ R.sumP $ R.map (renderPoint logMul maxV) vectorField
    image =
      runIdentity $ R.computeUnboxedP $
      R.map (\v ->
               let
                 V3 r g b = fmap (round . (*255) . (/aaSq')) v
               in (r,g,b)
            ) image'
  in
    maxV `seq` image' `deepSeqArray` image

renderPoint
  :: (Epsilon a, RealFrac a, Floating a, Ord a)
  => Maybe a --logMul
  -> a --maxV
  -> V2 a
  -> V3 a
renderPoint mlogMul maxV v@(V2 y _x) =
  let
    theta | y < 0 = 2*pi - theta'
          | otherwise = theta'
      where theta' = acos $ (V2 0 1) `dot` normalize v
    applyLogFilter n =
      case mlogMul of
        Nothing -> n
        Just logMul -> log $ (n*logMul + 1)
    h' = 360*theta/(2*pi)
    s' = 0.5
    v' = (*0.6) $ applyLogFilter (norm v) / applyLogFilter maxV
  in uncurryRGB V3 $ hsl h' s' v'


src/Field.hs
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}

module Field where

import Control.Lens
import Control.Monad.Identity (runIdentity)

import Data.Array.Repa

import Data.Vector.Unboxed (Vector)
import Data.Vector.Unboxed as UV

import Linear

data FieldDescription a =
  FromCenter
  { _res :: !(V2 Int)
  , _center :: !(V2 a)
  , _h :: !a
  , _aa :: Int
  } deriving Show

defaultFieldDescription :: Num a => FieldDescription a
defaultFieldDescription = FromCenter (V2 1080 1920) (V2 0 0) 1 1

v2ToTuple :: Getter (V2 a) (a,a)
v2ToTuple = to (\(V2 y x) -> (x,y))

tupleToV2 :: Getter (a,a) (V2 a)
tupleToV2 = to (\(x,y) -> V2 y x)

makeLenses ''FieldDescription

generateCoords :: Fractional a => FieldDescription a -> (V3 Int -> V2 a)
generateCoords fd =
  let
    h' = _h fd
    res'@(V2 resY' resX') = fmap fromIntegral $ _res fd
    w = let aspect = resX' / resY' in aspect * h'
    offset' = _center fd - V2 h' w
    multiplier' = 2 * V2 h' w / res'
    aa' = _aa fd
    aaSq = aa'*aa'
    aaSq' = fromIntegral aaSq
  in
    (\(V3 y x i) ->
       let
         aaoffset =
           fmap ((/aaSq') . fromIntegral) $ V2 (i `div` aa') (i `rem` aa')
       in
         multiplier' * (fmap fromIntegral (V2 y x) + aaoffset) + offset'
    )

{-# INLINABLE generateCoords #-}
{-# SPECIALIZE generateCoords :: FieldDescription Float -> (V3 Int -> V2 Float) #-}
{-# SPECIALIZE generateCoords :: FieldDescription Double -> (V3 Int -> V2 Double) #-}

buildFieldRepa
  :: forall a b. (Fractional a, Unbox b)
  => FieldDescription a
  -> (V2 a -> V2 b)
  -> Array U DIM3 (V2 b)
buildFieldRepa fd fg =
  let
    V2 resY resX = _res fd
    coordinator = generateCoords fd
    aa' = _aa fd
    dim = Z :. resY :. resX :. aa'*aa'
  in
    runIdentity $ computeUnboxedP $ fromFunction dim
    (\(Z :. y :. x :. i) -> fg . coordinator $ (V3 y x i))

buildField
  :: forall a b. (Unbox b, Fractional a)
  => FieldDescription a
  -> (V2 a -> V2 b)
  -> Vector (V2 b)
buildField fd fg =
  let
    aa' = _aa fd
    aaSq = aa'*aa'
    coordinator :: Int -> V2 a
    coordinator i =
      generateCoords fd $
      V3 (i `div` (resX*aaSq)) (i `rem` (resX*aaSq) `div` aaSq) (i `rem` aaSq)
    V2 resY resX = _res fd
  in
    UV.generate (resY*resX*aaSq) $ fg . coordinator


scripts/generate-examples.sh
#!/bin/bash

# Simple limit cycle
time stack exec -- two-danode-exe -P -R "r*(1-r^2)" -T "1" -H 1.5 -c "(0,0)" -m 20 -r "(4096,4096)" -f 001.bmp

# Spiral with homoclinic orbit
time stack exec -- two-danode-exe -F "y-y^3" -G "-x-y^2" -H 2 -c "(0,0)" -m 20 -r "(4096,4096)" -f 002.bmp

time stack exec -- two-danode-exe -F "-2*cos x - cos y" -G "-2*cos y - cos x" -H 4 -c "(0,0)" -m 5 -r "(4096,4096)" -f 003.bmp

# Pendulum
time stack exec -- two-danode-exe -F "y" -G "-sin x" -H 8 -c "(0,0.0)" -m 10 -r "(4096,4096)" -f 004.bmp

# Damped pendulum
time stack exec -- two-danode-exe -F "y" -G "-1*y-sin x" -H 8 -c "(0,0.0)" -m 10 -r "(4096,4096)" -f 005.bmp

# Van der pol oscillator
time stack exec -- two-danode-exe -F "y" -G "-x+y*(1-x^2)" -H 4 -c "(0,0)" -m 10 -r "(4096,4096)" -f 006.bmp

time stack exec -- two-danode-exe -F "y^3 - 4*x" -G "y^3 - y - 3*x" -H 6 -c "(0,0)" -m 1 -r "(4096,4096)" -f 007.bmp
time stack exec -- two-danode-exe -F "y^3 - 12*x" -G "y^3 - y - 10*x" -H 3 -c "(0,0)" -m 10 -r "(4096,4096)" -f 008.bmp
time stack exec -- two-danode-exe -F "-x+1*y+x*x*y" -G "1-1*y-x*x*y" -H 2 -c "(0,0.5)" -m 100 -r "(4096,4096)" -f 009.bmp

# Infinite-period bifurcations
time stack exec -- two-danode-exe -R "r*(1-r*r)" -T "1 - sin theta" -P -H 2 -m 10 -r "(4096,4096)" -f 010.bmp
time stack exec -- two-danode-exe -R "r*(1-r*r)" -T "0.5 - sin theta" -P -H 2 -m 20 -r "(4096,4096)" -f 011.bmp

time stack exec -- two-danode-exe -F "1 - (1+1)*x+1*x*x*y" -G "1*x-1*x*x*y" -H 5 -m 20 -r "(4096,4096)" -f 012.bmp

# Competition models for shared resource
time stack exec -- two-danode-exe -F "x*(1-x/100) - x*y" -G "y*(1-y/20)-x*y" -H 2 -m 10 -r "(4096,4096)" -f 013.bmp
time stack exec -- two-danode-exe -F "x*(1-x/10) - x*y" -G "y*(1-y/5)-x*y" -H 5 -r "(4096,4096)" -f 014.bmp
time stack exec -- two-danode-exe -F "1.25*x*(1-x/10) - x*y" -G "y*(1-y/5)-x*y" -H 5 -m 10 -r "(4096,4096)" -f 015.bmp

# Predator prey model
time stack exec -- two-danode-exe -F "0.75*x*(1-x/2) - 2*x*y/(1+x)" -G "-y + 2*x*y/(1+x)" -H 0.75 -c "(0.75,0.75)" -m 20 -r "(4096,4096)" -f 016.bmp

time stack exec -- two-danode-exe -P -R "2*(1-2*sin theta + r * (1-r*r*r))" -T "1 - 0.5*cos theta + sin theta * r" -H 1.5 -c "(0,-0.25)" -m 8 -r "(4096,4096)" -f 017.bmp
time stack exec -- two-danode-exe -P -c "(0,0)" -R "r*(1-r)*(r-1)*(0.5-r)" -T "sin (12*theta)" -H 2.2  -m 300 -r "(4096,4096)" -f 018.bmp
time stack exec -- two-danode-exe -P -c "(0,0)" -m 300 -R "r*(r-1)*(0.5-r)" -T "sin (12*theta)*(cos (8*theta))" -H 2 -r "(4096,4096)" -f 019.bmp
time stack exec -- two-danode-exe -F "(cos (5*x))*(1-x)*(x-1)*(0.5-x)" -G "20*cos (1-x)*(sin (5*y))*(1-y)" -H 5 -m 30 -r "(4096,4096)" -f 020.bmp
time stack exec -- two-danode-exe -F "(cos (5*x*y))*(1-x*y)*sin (10*(x*y-1))*(0.5-x*y)" -G "0.1*(0.5-x*y)*10*cos (1-x*y*y)*(sin (5*y*x))*(1-y*x)" -H 4 -m 3000 -r "(4096,4096)" -f 021.bmp
time stack exec -- two-danode-exe -F "(cos (5*x*y))**sin (10*(x*y-1))*(0.5-x*y)" -G "0.1*(0.5-x*y)*10*cos (2-x*y*y)*(sin (5*y*x))" -H 4 -m 3000 -r "(4096,4096)" -f 022.bmp
time stack exec -- two-danode-exe -F "5*(cos (5*x*y))**sin (10*(x*y-1))*(0.5-x*y)" -G "0.1*(0.5-x*y)*10*cos (2-x*y*y)*(sin (5*y*x))" -H 4 -m 3000 -r "(4096,4096)" -f 023.bmp
time stack exec -- two-danode-exe -F "2*exp(cos (10*x) - sin (30*y))*sin(0.5 - 10*x*y)" -G "2*exp(sin (20*x) - cos(30*y))*sin(1 + 10*x*y)" -H 1 -m 3 -r "(4096,4096)" -f 024.bmp
time stack exec -- two-danode-exe -F "2*exp(cos (10*x*y) - sin (30*y*x))*sin(0.5 - 10*x*y)" -G "2*exp(sin (20*x) - cos(30*x*y))*sin(1 + 10*x*y)" -H 1 -m 3 -r "(4096,4096)" -f 025.bmp
time stack exec -- two-danode-exe -F "2*exp(cos (10*x*y) - sin (30*y*x))*sin(0.5 - 10*x*y)" -G "2*exp(sin (20*sqrt (abs (x*x*y))) - cos(30*x*y))*sin(1 + 10*x*y)" -H 1.3 -m 3 -r "(4096,4096)" -f 026.bmp
time stack exec -- two-danode-exe -F "x" -G "y" -H 8 -c "(0,0.0)" -m 10 -r "(4096,4096)" -f 027.bmp
time stack exec -- two-danode-exe -F "-2*cos x - cos y" -G "-2*cos y - cos x" -H 4 -c "(0,0)" -m 0.1 -r "(4096,4096)" -f 028.bmp
time stack exec -- two-danode-exe -F "y-y^3" -G "-x-y^2" -H 2 -c "(0,0)" -m 3 -r "(4096,4096)" -f 029.bmp
time stack exec -- two-danode-exe -F "-2*cos x - cos y" -G "-2*cos y - cos x" -H 4 -c "(0,0)" -r "(4096,4096)" -f 030.bmp
time stack exec -- two-danode-exe -F "2*exp(cos (10*x*y) - sin (30*y*x))*sin(0.5 - 10*x*y)" -G "2*exp(sin (20*sqrt (abs (x*x*y))) - cos(30*x*y))*sin(1 + 10*x*y)" -H 1.5 -r "(4096,4096)" -f 031.bmp
time stack exec -- two-danode-exe -H 5 -F "sin x - cos (2.3*y)" -G "cos y - sin (4*x)" -r "(4096,4096)" -f 032.bmp
time stack exec -- two-danode-exe -P -R "2*(1-2*sin theta + r * (1-r*r*r))" -T "1 - 0.5*cos theta + sin theta * r" -H 1.5 -c "(0,-0.25)" -r "(4096,4096)" -f 033.bmp
time stack exec -- two-danode-exe -P -c "(0,0)" -R "r*(1-r)*(r-1)*(0.5-r)" -T "sin (12*theta)" -H 2.2 -r "(4096,4096)" -f 034.bmp
time stack exec -- two-danode-exe -P -c "(0,0)" -R "r*(r-1)*(0.5-r)" -T "sin (12*theta)*(cos (8*theta))" -H 2 -r "(4096,4096)" -f 035.bmp
time stack exec -- two-danode-exe -F "(cos (5*x))*(1-x)*(x-1)*(0.5-x)" -G "20*cos (1-x)*(sin (5*y))*(1-y)" -H 5 -r "(4096,4096)" -f 036.bmp
time stack exec -- two-danode-exe -F "(cos (5*x*y))*(1-x*y)*sin (10*(x*y-1))*(0.5-x*y)" -G "0.1*(0.5-x*y)*10*cos (1-x*y*y)*(sin (5*y*x))*(1-y*x)" -H 4 -r "(4096,4096)" -f 037.bmp
time stack exec -- two-danode-exe -F "(cos (5*x*y))**sin (10*(x*y-1))*(0.5-x*y)" -G "0.1*(0.5-x*y)*10*cos (2-x*y*y)*(sin (5*y*x))" -H 4 -r "(4096,4096)" -f 038.bmp
time stack exec -- two-danode-exe -F "5*(cos (5*x*y))**sin (10*(x*y-1))*(0.5-x*y)" -G "0.1*(0.5-x*y)*10*cos (2-x*y*y)*(sin (5*y*x))" -H 4 -r "(4096,4096)" -f 039.bmp
time stack exec -- two-danode-exe -F "2*exp(cos (10*x) - sin (30*y))*sin(0.5 - 10*x*y)" -G "2*exp(sin (20*x) - cos(30*y))*sin(1 + 10*x*y)" -H 1 -r "(4096,4096)" -f 040.bmp
time stack exec -- two-danode-exe -F "2*exp(cos (10*x*y) - sin (30*y*x))*sin(0.5 - 10*x*y)" -G "2*exp(sin (20*x) - cos(30*x*y))*sin(1 + 10*x*y)" -H 1 -r "(4096,4096)" -f 041.bmp
time stack exec -- two-danode-exe -F "2*exp(cos (10*x*y) - sin (30*y*x))*sin(0.5 - 10*x*y)" -G "2*exp(sin (20*sqrt (abs (x*x*y))) - cos(30*x*y))*sin(1 + 10*x*y)" -H 1.3 -r "(4096,4096)" -f 042.bmp
time stack exec -- two-danode-exe -H 5 -F "(abs $ cos (0.75*x))**(abs $ cos (3*y)) + sin (3*x)" -G "sin x - cos y" -r "(4096,4096)" -f 043.bmp
time stack exec -- two-danode-exe -H 5 -F "(1.5*sin (3*x*y -10))*(abs $ cos (1.75*x))**(abs $ 3 * cos (3*y)) + sin (3*x)" -G "sin (4.3*x) - cos (0.5*y)" -r "(4096,4096)" -f 044.bmp
time stack exec -- two-danode-exe -H 8 -F "(1.5*sin (3*x*y -10))*(abs $ cos (1.75*x))**(abs $ 3 * cos (3*y)) + sin (3*x)" -G "sin (4.3*x) - cos (0.5*y)" -r "(4096,4096)" -f 045.bmp
time stack exec -- two-danode-exe -H 5 -r "(4096,4096)" -f 046.bmp -F "sin(2*y - 3*x)*exp(-abs (sin (3*x+1-2*y) - cos(x-3*y+1)))" -G "exp(-abs (3*cos y - 2*sin (1*x) + 3*cos(x*y) - 2.5*sin(0.5*x*y)))"
time stack exec -- two-danode-exe -H 1 -r "(4096,4096)" -f 047.bmp -F "sin(2*y - 3*x)*exp(-abs (sin (3*x+1-2*y) - cos(x-3*y+1)))" -G "cos(2*y - 3*x)*exp(-abs (cos (3*x+1-2*y) - sin(x-3*y+1)))"
time stack exec -- two-danode-exe -H 3 -r "(4096,4096)" -f 048.bmp -F "sin(2*y - 3*x)*exp(-abs (sin (3*x+1-2*y) - cos(x-3*y+1)))" -G "cos(2*y - 3*x)*exp(-abs (cos (3*x+1-2*y) - sin(x-3*y+1)))"
time stack exec -- two-danode-exe -H 5 -r "(4096,4096)" -f 049.bmp -F "sin(2*y - 3*x)*exp(-abs (sin (3*x+1-2*y) - cos(x-3*y+1)))" -G "cos(2*y - 3*x)*exp(-abs (cos (3*x+1-2*y) - sin(x-3*y+1)))"
time stack exec -- two-danode-exe -H 3 -r "(4096,4096)" -f 050.bmp -F "sin(2*y - 3*x)*exp(-abs (sin (3*x+1-2*y) - cos(x-3*y+1)))" -G "cos(2*y - 3*x)*exp(-abs (cos (3*x+1-2*y) - cos(x-3*y+1)))"
time stack exec -- two-danode-exe -H 5 -r "(4096,4096)" -f 051.bmp -F "sin(2*y - 3*x)*exp(-abs (sin (3*x+1-2*y) - cos(x-3*y+1)))" -G "cos(2*y - 3*x)*exp(-abs (cos (3*x+1-2*y) - cos(x-3*y+1)))"
time stack exec -- two-danode-exe -H 5 -r "(4096,4096)" -f 052.bmp -F "cos(2*y - 3*x)*exp(-abs (cos (3*x+1-2*y) - cos(x-3*y+1)))" -G "cos(2*y - 3*x)*exp(-abs (cos (3*x+1-2*y) - cos(x-3*y+1)))"

#More Competition models for shared resources
time stack exec -- two-danode-exe -F "x*(3-x - 2*y)" -G "y*(2-x-y)" -H 4 -m 10 -r "(4096,4096)" -f 053.bmp
time stack exec -- two-danode-exe -F "x*(1-x/7) - x*y" -G "y*(1-y/4)-x*y" -H 10 -m 10 -r "(4096,4096)" -f 054.bmp
